src/tests/tests.rs

```diff
@@ -476,3 +477,207 @@ fn production_mode_enclave_attest_works_when_sgx_debug_mode_not_allowed() {
         //TODO
     })
 }
+
+#[test]
+fn verify_unshield_funds_works() {
+    new_test_ext().execute_with(|| {
+        Timestamp::set_timestamp(TEST4_TIMESTAMP);
+        let signer4 = get_signer(TEST4_SIGNER_PUB);
+        let call_hash: H256 = H256::from([1u8; 32]);
+        let bonding_account = get_signer(&TEST4_MRENCLAVE);
+        let incognito_account = INCOGNITO_ACCOUNT.to_vec();
+
+        //Register enclave
+        assert_ok!(Teerex::register_enclave(
+            Origin::signed(signer4.clone()),
+            TEST4_CERT.to_vec(),
+            URL.to_vec(),
+        ));
+        assert_eq!(Teerex::enclave_count(), 1);
+
+        assert!(Teerex::shield_funds(
+            Origin::signed(AccountKeyring::Alice.to_account_id()),
+            incognito_account.clone(),
+            100,
+            bonding_account.clone(), //Bonding account (Eoc)
+        )
+        .is_ok());
+
+        assert_eq!(Balances::free_balance(bonding_account.clone()), 100);
+
+        let expected_event = Event::Teerex(RawEvent::ShieldFunds(incognito_account));
+        assert!(System::events().iter().any(|a| a.event == expected_event));
+
+        assert!(Teerex::unshield_funds(
+            Origin::signed(signer4.clone()), //Alice incognito account
+            AccountKeyring::Alice.to_account_id(),
+            50,
+            bonding_account.clone(),
+            call_hash.clone()
+        )
+        .is_ok());
+        assert_eq!(Balances::free_balance(bonding_account), 50);
+
+        let expected_event = Event::Teerex(RawEvent::UnshieldedFunds(
+            AccountKeyring::Alice.to_account_id(),
+        ));
+        assert!(System::events().iter().any(|a| a.event == expected_event));
+    })
+}
+
+#[test]
+fn verify_unshield_funds_from_not_registered_enclave_fails() {
+    new_test_ext().execute_with(|| {
+        Timestamp::set_timestamp(TEST4_TIMESTAMP);
+        let signer4 = get_signer(TEST4_SIGNER_PUB);
+        let call_hash: H256 = H256::from([1u8; 32]);
+
+        assert_eq!(Teerex::enclave_count(), 0);
+
+        assert!(Teerex::unshield_funds(
+            Origin::signed(signer4.clone()),
+            AccountKeyring::Alice.to_account_id(),
+            51,
+            signer4.clone(),
+            call_hash.clone()
+        )
+        .is_err());
```

Comment:
I we have return error variants in the pallet, we can also check for the specific error here like we do here:

https://github.com/integritee-network/pallet-teerex/blob/3124348b3be86303f7b83ca159bf9353dcb5306f/src/tests/tests.rs#L185